### PR TITLE
Prep work for second GeRM run

### PR DESCRIPTION
**Pull request by tacaswell**

### CODE REVIEW

**Comment by tacaswell** [comment]
To compile

```
g++ germ_udpsrvr.c -fpermissive -lzmq -o udp_collector
```

To test, open two terminals, in one run `python cli/det_sim.py` (first) and then in another run `./udp_collector`.

Then open a third terminal with IPython

```python
import zmq
ct = zmq.Context()
fname_sock = ct.socket(zmq.REQ)
go_sock = ct.socket(zmq.REQ)

go_sock.connect('tcp://localhost:5555')
fname_sock.connect('tcp://localhost:5557')
```

will set up the zmq sockets needed to drive things.

To drive a data-collection cycle:

```python
# sets the file name, will have `_NNN.bin` appended
fname_sock.send(b'/tmp/test'); fname_sock.recv()
# triggers data collection
go_sock.send(struct.pack('III', 1, 0, 1)); go_sock.recv()
# tells the udp collector to actually write the file, the recv blocks till done
fname_sock.send(b'ack'); fname_sock.recv()
# acknowledge that file is written
fname_sock.send(b'ack'); fname_sock.recv()
```

**Comment by tacaswell** [comment]
The work that is still left is to encode this extra logic into the IOC so `germ.trigger()` works correctly from bluesky.

**Comment by tacaswell** [comment]
@jrmlhermitte You should be able to push to this branch on my repository.

File: cli/det_sim.py [hunk 1]
```diff
@@ -80,43 +127,83 @@ def recv_and_process():
...
+                udp.send(b''.join((header,
+                                   bytes(head))))
+                udp_packet_count += 1
+            else:
+                first_head = 511 - len(tail)
+                head = (tail, payload[:first_head])
+                tail = payload[first_head:]
+                header = struct.pack('!Ixxxx', udp_packet_count)
+                udp.send(b''.join((header,
+                                   bytes(head[0]),
```
Comment (jrmlhermitte) [inline cli/det_sim.py]: these are all `np.ndarrays` for me and had to change to `head[0].tobytes()` etc.
do you know why this worked for you? I have numpy 1.14
(If I didn't do this I got an error. It works fine now)

Comment (tacaswell) [reply cli/det_sim.py]: What version of python?  There was a subtle bug around this sort of thing that we hit in caproto (@danielballan and @klauer may remember the exact details better than I do).

Comment (jrmlhermitte) [reply cli/det_sim.py]: I'm using 3.6.0. Not sure what it is. Is it safer to call `.tobytes` anyway?

Comment (klauer) [reply cli/det_sim.py]: Reference: https://github.com/NSLS-II/caproto/issues/48

Comment (tacaswell) [reply cli/det_sim.py]: update to 3.6.2 and that should be fixed.

**Comment by tacaswell** [comment]
Start the IOC via (you will need curio + caproto master installed)

```
python germ_ioc_udp.py localhost localhost
```

To test this with ophyd/bluesky

```python
from pygerm.ophyd import GeRM
g = GeRM('XF:28IDC-ES:1{Det:GeRM1}', name='g')
```

doing `g.trigger()` should produce spam in all three terminals.

For those keeping score at home we now need 4 processes (det_sim, udp_collector, germ_ioc_udp, IPython) to test this :smiling_imp: .

**Comment by tacaswell** [comment]
Still todo

 - [x] make the path configurable
 - [x] make sure waiting works the way I expect
 - [x] deal with asset registry insert
 - [x] get udp collector to return frame number, total events, overflows over zmq channel

File: pygerm/handler.py [hunk 2]
```diff
@@ -40,7 +40,7 @@ def __init__(self, fpath):
             raise ValueError(msg)
 
         # remove first and last region
-        raw_data = raw_data[2:-1]
+        raw_data = raw_data[2:-2]
```
Comment (jrmlhermitte) [inline pygerm/handler.py]: oh yes, thanks for catching that!

**Comment by tacaswell** [comment]
I'm going to merge this so we can start to deploy from the upstream branch not my branch.

**Pull request merged**